poseidon2/io.nim

```diff
@@ -3,32 +3,35 @@ import constantine/math/arithmetic
 import constantine/math/io/io_bigints
 import constantine/math/config/curves
 
-func fromBytes*(_: type F, bytes: openArray[byte]): F =
+export curves
+
+func fromOpenArray(_: type F, bytes: openArray[byte]): F =
+  F.fromBig(B.unmarshal(bytes, littleEndian))
+
+func fromBytes*(_: type F, bytes: array[31, byte]): F =
   ## Converts bytes into a field element. The byte array is interpreted as a
-  ## canonical little-endian big integer. The array should be of length 31 bytes
-  ## or less to ensure that it fits in a field of 254 bits. The remaining
-  ## 6 most-significant bits are set to 0.
-  assert bytes.len <= 31
-  var padded: array[32, byte]
-  copyMem(addr padded[0], unsafeAddr bytes[0], bytes.len)
-  let bigint = B.unmarshal(padded, littleEndian)
-  return F.fromBig(bigint)
+  ## canonical little-endian big integer.
+  F.fromOpenArray(bytes)
+
+func toBytes*(element: F): array[32, byte] =
+  ## Converts a field element into its canonical representation in little-endian
+  ## byte order. Uses at most 254 bits, the remaining most-significant bits are
+  ## set to 0.
+  assert marshal(result, element.toBig(), littleEndian)
 
-func fromBytes*(_: type seq[F], bytes: openArray[byte]): seq[F] =
+iterator elements*(bytes: openArray[byte], _: type F): F =
   ## Converts bytes into field elements. The byte array is converted 31 bytes at
-  ## a time with the `F.fromBytes()` function.
+  ## a time with the `F.fromBytes()` function. An end marker (0x1) and
+  ## padding (0x0's) are added to ensure unique field elements for byte
+  ## sequences that end with 0x0's.
   const chunkLen = 31
-  var elements: seq[F]
+  const endMarker = @[1'u8]
   var chunkStart = 0
-  while chunkStart < bytes.len:
-    let chunkEnd = min(chunkStart + 31, bytes.len)
-    let element = F.fromBytes(bytes.toOpenArray(chunkStart, chunkEnd - 1))
-    elements.add(element)
+  while chunkStart + chunkLen <= bytes.len:
+    let chunkEnd = chunkStart + chunkLen - 1
+    let element = F.fromOpenArray(bytes.toOpenArray(chunkStart, chunkEnd))
```

Comment:
I think you mean to call `fromBytes` here instead of `F.fromOpenArray`? 

(they are the same, but it's better for readability, and looks safer in case of future changes. Also that's what the comment claims :)

Reply:
`fromBytes` expects an `array`, that won't work, unless there is another override that takes an open array.